bench: Initial wasm benchmarks

simar7 opened the pull request:
<img width="715" alt="image" src="https://user-images.githubusercontent.com/1254783/123492739-363f5980-d5cf-11eb-96f7-fd8fe204c166.png">



```
➜  benchmark git:(wasm-benchmarks) ✗ go test -bench=. -benchmem -benchtime=100x  | prettybench -no-passthrough
benchmark                                     iter           time/iter     bytes alloc             allocs
---------                                     ----           ---------     -----------             ------
BenchmarkWasm_SimpleRule-8                     100      13646.00 ns/op       1684 B/op       30 allocs/op
BenchmarkRegoGoLibrary_SimpleRule-8            100      16221.00 ns/op       8040 B/op      123 allocs/op
BenchmarkWasm_RealRule-8                       100      30889.00 ns/op       1992 B/op       30 allocs/op
BenchmarkRegoGoLibrary_RealRule-8              100      40142.00 ns/op      17720 B/op      365 allocs/op
BenchmarkOnEventCodeInjectionRuleRego-8        100      40350.00 ns/op      18022 B/op      367 allocs/op
BenchmarkOnEventCodeInjectionRuleGo-8          100        814.90 ns/op        560 B/op        4 allocs/op
BenchmarkOnEventCodeInjectionRuleWASM-8        100      24217.00 ns/op       1912 B/op       30 allocs/op
BenchmarkEngineWithCodeInjectionRuleRego-8     100   13143482.00 ns/op    5483945 B/op   116421 allocs/op
BenchmarkEngineWithCodeInjectionRuleGo-8       100    1264845.00 ns/op     210097 B/op     2008 allocs/op
BenchmarkEngineWithCodeInjectionRuleWASM-8     100   30494157.00 ns/op    1725378 B/op    25678 allocs/op
BenchmarkEngineWithMultipleRulesRegoAndGo-8    100   16792141.00 ns/op   10575072 B/op   224289 allocs/op
```

Signed-off-by: Simar <simar@linux.com>

danielpacak commented on tracee-rules/engine/engine.go:
```diff
@@ -69,7 +69,6 @@ func NewEngine(sigs []types.Signature, sources EventSources, output chan types.F
 func (engine *Engine) Start(done chan bool) {
 	go engine.consumeSources(done)
 	for s, c := range engine.signatures {
-		defer close(c)
```
Let's do that in a dedicated PR after we reach the consensus whether to close or not to close the channel. This change deserves its own commit with a bit of explanation and I don't think we want to smuggle it with benchmarks. What's more we have #777 to track it.

For sure we have to do sth about it as I'm also seeing this warning in my IDEA: **Possible resource leak, 'defer' is called in the 'for' loop**

simar7 replied:
Yeah sorry I didn't mean to include this change in this PR. I'll revert it back.

danielpacak commented on tracee-rules/main.go:
```diff
@@ -38,10 +40,21 @@ func main() {
-			if err != nil {
-				return err
-			}
+			//sigs, err := getSignatures(c.String("rules-dir"), c.StringSlice("rules"))
```
I believe we should revert that?

simar7 replied:
Yeah more cruft I'll undo this.

danielpacak commented on tracee-rules/benchmark/benchmark_reference_test.go:
```diff
@@ -0,0 +1,160 @@
+// Reference: https://github.com/open-policy-agent/golang-opa-wasm/blob/8041bec28e319aad31fd4632bae46e47727f3ddc/opa/opa_test.go#L154-L216
```
What is the difference between benchmark_reference_test.go and benchmark_test.go? Do we need both Go test files?

simar7 replied:
`benchmark_reference_test.go` is a mostly standalone benchmark that shows performance between WASM/Rego without any additional tracee dependencies. I added it in so as to share with the OPA team if they need to evaluate it. 

We can put them in the same file if that's better.

itaysk replied:
The goal is to have reliable and useful benchmarks that can help us make decisions, sharing them externally is desirable side-effect but shouldn't drive the design of the tests. To address the specific points:
- I don't see a point in benchmarking both sample signatures and a real signatures. The content and name of the signature is meaningless but it needs to so something typical like 3-5 lines of assertions on input at least. Any of the existing signatures would be ok.
- We shouldn't use or test the "golang-opa-wasm" library, as suggested by the maintainers
- I agree it's better to make changes in the upstream code as opposed to rewriting the tested code in the test
- However - we shouldn't automatically upstream all the possible variations we benchmark (that's why we test them first). For specific cases, like WASM, sounds like a quick win we would want to have anyway, so yeah. For other cases, maybe we can consider upstreaming support for user-provided flags if we want it to be more generic and not necessarily in response to a benchmark (won't help all variations though).
- If we want to keep the "reference" benchmarks just for sharing externally, we could move it to personal repo (they don't really relate to tracee)

danielpacak reviewed (changes requested):
Hey @simar7 I left a few comments but the main thing is to see if we can use opa.Target option instead of github.com/open-policy-agent/golang-opa-wasm module. By doing that we could benchmark same Rego rules with different targets (rego vs wasm)

simar7 commented on tracee-rules/benchmark/signature/wasm/signatures.go:
```diff
@@ -0,0 +1,202 @@
+	//fmt.Println(pkgName)
+
+	cr, err := rego.New(
+		rego.Compiler(compiledRego),
```
I tried it but the results are rather unexpectedly slower.

```
 go test -tags=opa_wasm -bench=. -benchmem -benchtime=100x | prettybench -no-passthrough
benchmark                                     iter           time/iter     bytes alloc             allocs
---------                                     ----           ---------     -----------             ------
BenchmarkWASMOptionTarget_SimpleRule-2         100      80832.00 ns/op      13086 B/op      372 allocs/op
BenchmarkWASMLibrary_SimpleRule-2              100      11549.00 ns/op       1678 B/op       30 allocs/op
BenchmarkRegoGoLibrary_SimpleRule-2            100       9814.00 ns/op       7964 B/op      123 allocs/op
BenchmarkWASMOptionTarget_RealRule-2           100     122719.00 ns/op      25430 B/op      664 allocs/op
BenchmarkWASMLibrary_RealRule-2                100      20513.00 ns/op       1984 B/op       30 allocs/op
BenchmarkRegoGoLibrary_RealRule-2              100      38014.00 ns/op      18290 B/op      379 allocs/op
```

Notice the build flag of `opa_wasm` must be passed in for it to use the WASM runtime. By default it is disabled. Not sure why that is. 

Regardless I will still push these new tests for completeness.

simar7 replied:
Latest released OPA version (v0.29.4):
```
go test -tags=opa_wasm -bench=. -benchmem -benchtime=100x | prettybench -no-passthrough
benchmark                                     iter           time/iter     bytes alloc             allocs
---------                                     ----           ---------     -----------             ------
BenchmarkWASMOptionTarget_SimpleRule-2         100     316014.00 ns/op      34621 B/op     1251 allocs/op
BenchmarkWASMLibrary_SimpleRule-2              100       9334.00 ns/op       1677 B/op       30 allocs/op
BenchmarkRegoGoLibrary_SimpleRule-2            100       8391.00 ns/op       7908 B/op      122 allocs/op
BenchmarkWASMOptionTarget_RealRule-2           100     343139.00 ns/op      47059 B/op     1553 allocs/op
BenchmarkWASMLibrary_RealRule-2                100      23539.00 ns/op       1984 B/op       30 allocs/op
BenchmarkRegoGoLibrary_RealRule-2              100      30570.00 ns/op      18069 B/op      377 allocs/op
```

danielpacak replied:
I thought that we'll just add BenchmarkOnEventCodeInjectionRuleWASM and BenchmarkEngineWithCodeInjectionRuleWASM to the family:
BenchmarkOnEventCodeInjectionRuleRego
BenchmarkOnEventCodeInjectionRuleGo
BenchmarkEngineWithCodeInjectionRuleRego
BenchmarkEngineWithCodeInjectionRuleWASM

This will allow us to relate benchmarks that use the same (or comparable) rule and inputs

simar7 commented on tracee-rules/benchmark/signature/wasm/signatures.go:
```diff
@@ -0,0 +1,202 @@
+	rego *opa.OPA
+}
+
+func (w *AntiDebuggingWASMSignature) Init(cb types.SignatureHandler) error {
```
Good point - I implemented it here https://github.com/aquasecurity/tracee/pull/790/commits/fa686eb565b8a4e8f79447bc478d172aa110417c

simar7 commented:
I cleaned up and refactored tests as per feedback. Now we have table driven tests for each type along with subtests. 

```
go test -tags=opa_wasm -bench=. -benchmem -benchtime=10x  | prettybench -no-passthrough


benchmark                                                     iter             time/iter      bytes alloc               allocs
---------                                                     ----             ---------      -----------               ------
BenchmarkOnEventCodeInjectionRule/rego-8                        10        49014.00 ns/op       18886 B/op        382 allocs/op
BenchmarkOnEventCodeInjectionRule/golang-8                      10          843.80 ns/op         784 B/op          6 allocs/op
BenchmarkOnEventCodeInjectionRule/wasm-8                        10       355581.00 ns/op       47793 B/op       1548 allocs/op
BenchmarkEngineWithCodeInjection/rego-8                         10     14332547.00 ns/op     5783922 B/op     122534 allocs/op
BenchmarkEngineWithCodeInjection/golang-8                       10      1386668.00 ns/op      229213 B/op       2027 allocs/op
BenchmarkEngineWithCodeInjection/wasm-8                         10    509183955.00 ns/op    49738114 B/op    1636090 allocs/op
BenchmarkEngineWithMultipleRules/rego_and_golang-8              10     13236959.00 ns/op     5903897 B/op     121787 allocs/op
BenchmarkEngineWithMultipleRules/wasm_and_golang-8              10    602044118.00 ns/op    50179214 B/op    1642625 allocs/op
BenchmarkEngineWithMultipleRules/rego_and_wasm-8                10    546562624.00 ns/op    55432076 B/op    1760155 allocs/op
BenchmarkEngineWithMultipleRules/rego_and_golang_and_wasm-8     10    682285338.00 ns/op    55946471 B/op    1765420 allocs/op
BenchmarkEngineWithNSignatures/rego/2Signatures-8               10     25229206.00 ns/op    11297500 B/op     239605 allocs/op
BenchmarkEngineWithNSignatures/rego/4Signatures-8               10     62644894.00 ns/op    22705860 B/op     481871 allocs/op
BenchmarkEngineWithNSignatures/rego/8Signatures-8               10    132168955.00 ns/op    45258380 B/op     961040 allocs/op
BenchmarkEngineWithNSignatures/rego/16Signatures-8              10    247037004.00 ns/op    90983470 B/op    1932369 allocs/op
BenchmarkEngineWithNSignatures/rego/32Signatures-8              10    668971862.00 ns/op   181030222 B/op    3843827 allocs/op
BenchmarkEngineWithNSignatures/rego/64Signatures-8              10   1306239333.00 ns/op   366165934 B/op    7768299 allocs/op
BenchmarkEngineWithNSignatures/rego/128Signatures-8             10   1933996468.00 ns/op   739737259 B/op   15714503 allocs/op
BenchmarkEngineWithNSignatures/golang/2Signatures-8             10      2367042.00 ns/op      414360 B/op       3951 allocs/op
BenchmarkEngineWithNSignatures/golang/4Signatures-8             10      4178486.00 ns/op      837276 B/op       8080 allocs/op
BenchmarkEngineWithNSignatures/golang/8Signatures-8             10      7220736.00 ns/op     1658461 B/op      15904 allocs/op
BenchmarkEngineWithNSignatures/golang/16Signatures-8            10     12930749.00 ns/op     3320600 B/op      32101 allocs/op
BenchmarkEngineWithNSignatures/golang/32Signatures-8            10     26639562.00 ns/op     6648820 B/op      64310 allocs/op
BenchmarkEngineWithNSignatures/golang/64Signatures-8            10     48356388.00 ns/op    13005915 B/op     126428 allocs/op
BenchmarkEngineWithNSignatures/golang/128Signatures-8           10    117156119.00 ns/op    26090901 B/op     253538 allocs/op
BenchmarkEngineWithNSignatures/wasm/2Signatures-8               10   1095853289.00 ns/op    98904970 B/op    3227186 allocs/op
```

simar7 commented:
I also left some remarks on this PR: https://github.com/aquasecurity/tracee/pull/796 regarding memory profile for WASM https://github.com/aquasecurity/tracee/pull/796#issuecomment-871021296 and CPU profile https://github.com/aquasecurity/tracee/pull/796#issuecomment-871017267
